Ops/Expr/Evaluate.cs

```diff
@@ -2,24 +2,45 @@
 using static Token.E;
 
 // An expression evaluator, implementing using the visitor pattern
-public class ExprEvaluator : Visitor<int> {
-   public ExprEvaluator (Dictionary<string, int> dict) => mDict = dict;
-   Dictionary<string, int> mDict;
+public class ExprEvaluator : Visitor<double> {
+   public ExprEvaluator (Dictionary<string, double> dict) => mDict = dict;
+   Dictionary<string, double> mDict;
 
-   public override int Visit (NLiteral literal)
-      => int.Parse (literal.Value.Text);
+   public override double Visit (NLiteral literal)
+      => double.Parse (literal.Value.Text);
 
-   public override int Visit (NIdentifier identifier)
+   public override double Visit (NIdentifier identifier)
       => mDict[identifier.Name.Text];
 
-   public override int Visit (NUnary unary) {
-      int d = unary.Expr.Accept (this);
+   public override double Visit (NUnary unary) {
+      double d = unary.Expr.Accept (this);
       if (unary.Op.Kind == SUB) d = -d;
       return d;
    }
 
-   public override int Visit (NBinary binary) {
-      int a = binary.Left.Accept (this), b = binary.Right.Accept (this);
+   public override double Visit (NFnCall fn) {
+      var args = fn.Params.Select (x => x.Accept (this)).ToArray ();
+      return fn.Name.Text.ToLower () switch {
+         "sin" => Math.Sin (D2R (args[0])),
+         "cos" => Math.Cos (D2R (args[0])),
+         "tan" => Math.Tan (D2R (args[0])),
+         "log" => Math.Log (args[0]),
+         "exp" => Math.Exp (args[0]),
+         "asin" => R2D (Math.Asin (args[0])),
+         "acos" => R2D (Math.Acos (args[0])),
+         "atan" => R2D (Math.Atan (args[0])),
+         "atan2" => R2D (Math.Atan2 (args[0], args[1])),
+         "random" => Random.Shared.Next (),
+         "round" => args.Length == 1 ? Math.Round (args[0]) : Math.Round (args[0], (int)args[1]),
+         _ => throw new NotImplementedException ()

```

Comment:
Fine for now, but we need an extensible system once people start writing their own functions in Pascal. Besides, even this list of possible function should be obtained from some "Runtime Library DLL"...

Reply:
OK.  Hopefully part of another assignment :). Thanks.